lib/segment/src/common/mmap_ops.rs

```diff
@@ -0,0 +1,137 @@
+use std::fs::OpenOptions;
+use std::mem::size_of;
+use std::ops::DerefMut;
+use std::path::Path;
+use std::{mem, slice};
+
+use memmap2::{Mmap, MmapMut, MmapOptions};
+
+use crate::entry::entry_point::OperationResult;
+use crate::madvise;
+
+pub fn create_and_ensure_length(path: &Path, length: usize) -> OperationResult<()> {
+    let file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(path)?;
+
+    file.set_len(length as u64)?;
+    Ok(())
+}
+
+pub fn open_read_mmap(path: &Path) -> OperationResult<Mmap> {
+    let file = OpenOptions::new()
+        .read(true)
+        .write(false)
+        .append(true)
+        .create(true)
+        .open(path)?;
+
+    let mmap = unsafe { MmapOptions::new().map(&file)? };
+    madvise::madvise(&mmap, madvise::get_global())?;
+    Ok(mmap)
+}
+
+pub fn open_write_mmap(path: &Path) -> OperationResult<MmapMut> {
+    let file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(false)
+        .open(path)?;
+
+    let mmap = unsafe { MmapMut::map_mut(&file)? };
+    madvise::madvise(&mmap, madvise::get_global())?;
+    Ok(mmap)
+}
+
+pub fn transmute_from_u8<T>(data: &[u8]) -> &T {
+    debug_assert!(data.len() == size_of::<T>());
+    let ptr = data.as_ptr() as *const T;
+    unsafe { &*ptr }
+}
+
+pub fn transmute_from_u8_to_mut<'a, T>(data: &mut [u8]) -> &'a mut T {
+    debug_assert!(data.len() == size_of::<T>());
+    let ptr = data.as_ptr() as *mut T;
+    unsafe { &mut *ptr }
+}
+
+pub fn transmute_to_u8<T>(v: &T) -> &[u8] {
+    unsafe { std::slice::from_raw_parts(v as *const T as *const u8, mem::size_of_val(v)) }
+}
+
+pub fn transmute_from_u8_to_slice<T>(data: &[u8]) -> &[T] {
+    debug_assert!(data.len() % size_of::<T>() == 0);
+    let len = data.len() / size_of::<T>();
+    let ptr = data.as_ptr() as *const T;
+    unsafe { std::slice::from_raw_parts(ptr, len) }
+}
+
+pub fn transmute_from_u8_to_mut_slice<T>(data: &mut [u8]) -> &mut [T] {
+    debug_assert!(data.len() % size_of::<T>() == 0);
+    let len = data.len() / size_of::<T>();
+    let ptr = data.as_mut_ptr() as *mut T;
+    unsafe { std::slice::from_raw_parts_mut(ptr, len) }
+}
+
+/// Transmute a `&mut [u8]` and `&mut [T]` and share mutable reference
+///
+/// # Safety
+///
+/// This is unsafe because this creates a second mutable reference to the given `data`.
```

Comment:
In general I think you should never introduce aliasing. Requiring that two mutable references are mutably accessed at the same time is not sufficient AFAIK, because the compiler might rely on no-aliasing for optimizations making this UB (not completely sure whether this is currently done, but there have been attempts to introduce LLVM's `noalias` feature). The rust reference [says the following about this](https://doc.rust-lang.org/reference/behavior-considered-undefined.html):

> Breaking the [pointer aliasing rules](http://llvm.org/docs/LangRef.html#pointer-aliasing-rules). Box<T>, &mut T and &T follow LLVM’s scoped [noalias](http://llvm.org/docs/LangRef.html#noalias) model, except if the &T contains an [UnsafeCell<U>](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html). References and boxes must not be [dangling](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#dangling-pointers) while they are live. The exact liveness duration is not specified, but some bounds exist:

Reply:
Do you think the better approach would be to consume mmap into some kind of wrapper object that ensures that only one mutable reference to the mmap exists?